corehq/ex-submodules/casexml/apps/phone/models.py

```diff
@@ -354,6 +355,53 @@ def test_only_get_dependent_cases_on_phone(self):
         raise NotImplementedError()
 
 
+def save_synclog_to_sql(synclog_json_object):
+    synclog = synclog_to_sql_object(synclog_json_object)
+    synclog.save()
+
+
+def synclog_to_sql_object(synclog_json_object):
+    # synclog_json_object should be a SyncLog instance
+    # if synclog_json_object._id
+    synclog = None
+    if synclog_json_object._id:
+        synclog = SyncLogSQL.objects.filter(synclog_id=synclog_json_object._id).first()
+
+    is_new_synclog_sql = not synclog_json_object._id or not synclog
+
+    if is_new_synclog_sql:
+        synclog_id = synclog_json_object._id or uuid.uuid4().hex.lower()
+        synclog_json_object._id = synclog_id
+        synclog = SyncLogSQL(
+            domain=synclog_json_object.domain,
+            user_id=synclog_json_object.user_id,
+            synclog_id=synclog_id,
+            date=synclog_json_object.date,
+            previous_synclog_id=getattr(synclog_json_object, 'previous_synclog_id', None),
+            log_format=synclog_json_object.log_format,
+        )
+    synclog.doc = synclog_json_object.to_json()
+    return synclog
+
+
+class SyncLogSQL(models.Model):
+    domain = models.CharField(max_length=255, null=True, blank=True, default=None, db_index=True)
+    user_id = models.CharField(max_length=255, default=None, db_index=True)
+    date = models.DateTimeField(db_index=True, null=True, blank=True)
+    synclog_id = models.UUIDField(unique=True, primary_key=True)
```

Comment:
If this is the primary key it should be the first field (by convention)

Reply:
Is it possible to make this something other than UUID right now because this comes from couch during the couch to sql migration. After the migration is over it might be easier to replace this with another incremental UUID  (the only refs that need update is previous_synclog_id)

Re. default, the UUID manually when instantiating an object otherwise there would be an exception, so is a default still necessary?

Reply:
> Is it possible to make this something other than UUID right now because this comes from couch during the couch to sql migration

We don't need to migrate this. All read queries should do something like:

```python
def get_synclog(synclog_id):
    try:
        return SyncLogSQL.get(synclog_id) 
    except:
        return CouchSyncLog.get(synclog_id)
```

That way we can fallback to the old synclogs without an active migration.

All write queries should only write to SQL.

> Re. default, the UUID is manually set when instantiating an object otherwise there would be an exception, so is a default still necessary?

I don't see where the UUID is being set manually. Can you link me to that? I don't think there's any harm with setting a default

Reply:
Sorry, I could have been more clear.

I already implemented the lazy migration (querying from couch and SQL for read but not for write instead of onetime migration). My above response was to your suggestion of making synclog_id (UUID) a sequential ID instead of couch like ID for better query performance. 

The manual UUID setting is done [here](https://github.com/dimagi/commcare-hq/pull/19391/files#diff-a7f81fae5762e257cd3241a8f458496dR400). Also if you look at that function it relies on old couch synclog IDs which are not sequential. So it's tricky to replace them with new sequential IDs as part of the lazy migration since it would be difficult to keep track of correspondence of couch and SQL synclogs. 

Once the lazy migration is fully complete (after 60 days), we can add a new ID column that is sequential and do a one time migration to replace the current synclog_id (UUID that is not sequential, that came from couch), it should be possible to do this very efficiently since all the all calculations will partain to the synclog sql table itself.

Reply:
Couch IDs do use a sequential uuids http://docs.couchdb.org/en/2.1.1/config/misc.html?highlight=sequential#uuids/algorithm

Anyways I think that may end up being a later optimization, but we may want to default to uuid1 (instead of uuid4) since it will be less random.

>  Also if you look at that function it relies on old couch synclog IDs which are not sequential

From my understanding, SyncLogs are never updated after it's sent to the user. So all new writes are only going to SQL and we won't write any docs that are in couch to postgres. When a user syncs after this PR, it will look like:

1) User "syncs to server"
2) In user's request there is a synclog_id ABC.
3) This code looks in SQL for synclog ABC, but it's not there. Looks in Couch gets synclog
4) restore code figures out what cases/UCRs to send back to the user
5) a new synclog DEF is generated with what information is being sent to the user.
6) synclog DEF is saved to SQL
7) restore is sent to user.
8) On next sync, synclog DEF is sent from user to HQ and all subsequent syncs are only relying on SQL

Synclog ABC will never be stored in SQL so there shouldn't be any reason that we're restricted due to couch ids

Reply:
Ah, that's right. Switching to uuid1, I had the migration script playing in my head when I thought about this.